lib/persistentqueue/fastqueue.go

```diff
@@ -132,26 +154,45 @@ func (fq *FastQueue) GetInmemoryQueueLen() int {
 	return len(fq.ch)
 }
 
-// MustWriteBlock writes block to fq.
-func (fq *FastQueue) MustWriteBlock(block []byte) {
+// MustWriteBlockIgnoreDisabledPQ writes block to fq, persists data on disk even if persistent disabled by flag.
+// it's needed to gracefully stop service and do not lose data if remote storage is not available.
+func (fq *FastQueue) MustWriteBlockIgnoreDisabledPQ(block []byte) {
```

Comment:
Is this needed because vmagent can't provide feedback to the client about write success?

Reply:
It's needed to gracefully store in-memory state.

vmagent reads a block from queue into memory for sending it to the remote storage. 

 If queue is blocked during vmagent shutdown, it cannot return read block back to it and in-memory data may be lost. To mitigate this case, vmagent need a function to perform writing into blocked queue.

Reply:
I see. I just wondered whether we need to be so precocious. It is ok for writer to do its best to deliver data, and if it is not possible during gracefulPeriod - drop data on the floor.

I'm afraid that this intention to be nice to the data could actually result in user's confusion. For example, vmagent with disabled queue could be deployed with ephemeral PVC (because user disabled the queue and thinks of vmagent as stateles). And this attempt to dumping data on disk won't make any difference.

I think, this can be explained via docs, though.

Reply:
If vmagent gracefully quitted and stored some in-memory data to disk[not ephemeral pvc for some reason], then when vmagent start again, it will keep returning [errFullQueue](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/gh-2110/lib/persistentqueue/fastqueue.go#L177) as long as the file-based queue not drained.
Could that be a little confusing since users already turn off OnDiskQueue, but it still block the first few requests?

Reply:
I think it's fine, stored part is really small and most users don't receive this error. 

 Also, errFullQueue is expected error for disabled on-disk queue mode.